libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt

```diff
@@ -1612,10 +1614,15 @@ class MapboxNavigation @VisibleForTesting internal constructor(
         latestLegIndex = null
         if (result.navigationRoutes.isNotEmpty()) {
             routeScope.launch {
-                val refreshed = routeRefreshController.refresh(result.navigationRoutes)
+                val indicesSnapshot = currentIndicesSnapshotProvider.freezeAndGet()
+                val refreshed = routeRefreshController.refresh(
+                    result.navigationRoutes,
+                    indicesSnapshot.legIndex
+                )
+                currentIndicesSnapshotProvider.unfreeze()
```

Comment:
This block is invoked as soon as the `RoutesObserver` fires, and sice we never clear the last `RouteProgress` value from `CurrentIndicesSnapshotProvider`, it points to the _previous_ route data or is missing entirely.

This causes the first route refresh always using the default `routeGeometryIndex = null` which falls back to the old logic, because the `RouteProgress` is not yet available when `RoutesObserver` fires.

Additional scenarios I'm thinking about:
- we drive along the route, already on `leg_index = 1` and `geometry_index=10`
- we change a route to a brand new one (reroute could be the reason)
- we land in this block of code, which would use the cached progress update (there's no guarantee that a fresh one was already delivered when `RoutesObserver` fires) and create indices snapshot of `leg_index = 1` and `geometry_index=10`, even though we're on `leg_index = 0` and `geometry_index=0`

Another scenario:
- we freeze the indices when the routes are set
- initiate the refresh, however, the refresh controller starts off with a delay equal to `RouteRefreshOptions#intervalMillis`
- when the delay elapses, we make a refresh request with geometry indices from before the initial delay and end up refreshing an already traveled portion of the route

I see a couple of things we might need to adjust here, unless the above scenarios are already taken into account somewhere that I'm not noticing:
- we need to wait for the first progress update before starting the refresh process, maybe we can move this coroutine launch to an internal progress observer?
- we should clear the indices snapshot cache when a session is stopped, or something along these lines to ensure that we're not using an outdated information

Reply:
The above scenario with using the default/outdated indices snapshot is also reproducible in instrumentation tests - we try doing an old refresh request (because progres is not yet available), the test keeps waiting until we make a correct request during a following attempt, and eventually succeeds.

I also see logs like:
```
2022-08-16 12:03:55.213 27691-28070/com.mapbox.navigation.instrumentation_tests E/Mapbox: [nav-native]: Directions Service: RouteRefresh unable to parse response Request url:
    /directions-refresh/v1/mapbox/driving-traffic/route_response_route_refresh/1/0?access_t
2022-08-16 12:03:55.217 27691-28838/com.mapbox.navigation.instrumentation_tests E/Mapbox: [nav-sdk]: [RouteRefreshController] Route refresh error: Route refresh failed throwable=java.lang.Exception: Route refresh failed.
    message = Unable to parse response
    code = 0
    type = UNKNOWN
    requestId = 5
    legIndex = 0
``` 
which could be related because the initial incorrect request ends up in https://github.com/mapbox/mapbox-navigation-android/blob/28a0023fe9e0c551a674d71d6edc9547203f62fc/libtesting-ui/src/main/java/com/mapbox/navigation/testing/ui/http/MockWebServerRule.kt#L46-L53 and looks like Nav Native doesn't pass that error back correctly to platform, NN reports `unable to parse response` instead of correctly parsing the error body. cc @DenisPryt